apps/webapp/app/v3/marqs.server.ts

```diff
@@ -0,0 +1,906 @@
+
+const SemanticAttributes = {
+  QUEUE: "marqs.queue",
+  PARENT_QUEUE: "marqs.parentQueue",
+  MESSAGE_ID: "marqs.messageId",
+  CONCURRENCY_KEY: "marqs.concurrencyKey",
+};
+
+// TODO: heartbeats from the workers to ensure they're still alive
+
+/**
+ * MarQS - Modular Asynchronous Reliable Queueing System (pronounced "markus")
+ */
+export class MarQS {
+  private redis: Redis;
+  #requeueingWorkers: Array<AsyncWorker> = [];
+
+  constructor(private readonly options: MarQSOptions) {
+    this.redis = new Redis(options.redis);
+
+    // Spawn options.workers workers to requeue visible messages
+    this.#startRequeuingWorkers();
+
+    this.#registerCommands();
+  }
+
+  public async updateQueueConcurrency(
+    env: AuthenticatedEnvironment,
+    queue: string,
+    concurrency: number
+  ) {
+    return this.redis.set(
+      `${constants.ENV_PART}:${env.id}:${constants.QUEUE_PART}:${queue}:${constants.CONCURRENCY_LIMIT_PART}`,
+      concurrency
+    );
+  }
+
+  public async enqueueMessage(
+    env: AuthenticatedEnvironment,
+    queue: string,
+    messageId: string,
+    messageData: Record<string, unknown>,
+    concurrencyKey?: string
+  ) {
+    return await this.#trace(
+      "enqueueMessage",
+      async (span) => {
+        const messageQueue = `${constants.ENV_PART}:${env.id}:${constants.QUEUE_PART}:${queue}${
+          concurrencyKey ? `:${constants.CONCURRENCY_KEY_PART}:${concurrencyKey}` : ""
+        }`;
+
+        const timestamp = Date.now();
+
+        const parentQueue =
+          env.type === "DEVELOPMENT"
+            ? `${constants.ENV_PART}:${env.id}:${constants.SHARED_QUEUE}`
+            : constants.SHARED_QUEUE;
+
+        const messagePayload: MessagePayload = {
+          version: "1",
+          data: messageData,
+          queue: messageQueue,
+          concurrencyKey,
+          timestamp,
+          messageId,
+          parentQueue,
+        };
+
+        span.setAttributes({
+          [SemanticAttributes.QUEUE]: queue,
+          [SemanticAttributes.MESSAGE_ID]: messageId,
+          [SemanticAttributes.CONCURRENCY_KEY]: concurrencyKey,
+          [SemanticAttributes.PARENT_QUEUE]: parentQueue,
+        });
+
+        await this.#callEnqueueMessage(messagePayload);
+      },
+      { kind: SpanKind.PRODUCER, attributes: { ...attributesFromAuthenticatedEnv(env) } }
+    );
+  }
+
+  public async dequeueMessageInEnv(env: AuthenticatedEnvironment) {
+    return this.#trace(
+      "dequeueMessageInEnv",
+      async (span, abort) => {
+        const parentQueue =
+          env.type === "DEVELOPMENT"
+            ? `${constants.ENV_PART}:${env.id}:${constants.SHARED_QUEUE}`
+            : constants.SHARED_QUEUE;
+
+        // Read the parent queue for matching queues
+        const messageQueue = await this.#getRandomQueueFromParentQueue(
+          parentQueue,
+          (queue, score) => this.#calculateMessageQueueWeight(queue, score)
+        );
+
+        if (!messageQueue) {
+          abort();
```

Comment:
Would it be useful to add some logging before all of these aborts?

Reply:
No because it's going to be VERY common for dequeue to abort, and the whole point is to prevent unnecessary logging.